File: backend/app/importers/canonical.py
"""Format wlasny aplikacji - `fiszki/v1`.

To jest format docelowy: kazde inne zrodlo jest do niego sprowadzane. Jest
tez formatem, w ktorym najlatwiej dostarczyc gotowa talie z zewnatrz (np.
wygenerowana przez model jezykowy), bo nie wymaga zgadywania kolumn.

    {
      "format": "fiszki/v1",
      "deck": "Angielski B2",
      "default_note_type": "basic_reversed",
      "notes": [
        {
          "front": "ubiquitous",
          "back": "wszechobecny",
          "example": "Smartphones are ubiquitous these days.",
          "tags": ["b2", "przymiotnik"],
          "kind": "word",
          "note_type": "basic_reversed",
          "source_ref": "oxford-3000/ubiquitous"
        }
      ]
    }

Wymagane sa wylacznie `front` i `back`. Przyjmowana jest tez naga lista notatek
(bez koperty) - wtedy obowiazuja wartosci domyslne.
"""

import json
from typing import Any

from app.core.rendering import FIELD_BACK, FIELD_EXAMPLE, FIELD_FRONT
from app.importers.base import (
    TARGET_KIND,
    TARGET_TAGS,
    ImportError_,
    ParseResult,
    SourceRow,
)
from app.models import NoteType

FORMAT_ID = "fiszki/v1"

#: Klucz w JSON -> kolumna posrednia (nazwa widoczna pozniej w mapowaniu).
_KEYS = {
    "front": FIELD_FRONT,
    "back": FIELD_BACK,
    "example": FIELD_EXAMPLE,
    "tags": TARGET_TAGS,
    "kind": TARGET_KIND,
}


def _as_text(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, (list, tuple)):
        return ", ".join(str(v) for v in value)
    return str(value)


def parse(data: bytes, options: dict | None = None) -> ParseResult:
    try:
        payload = json.loads(data.decode("utf-8-sig"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ImportError_(f"Nieprawidlowy JSON: {exc}") from exc

    warnings: list[str] = []

    if isinstance(payload, list):
        notes = payload
        deck_name = None
        default_note_type = NoteType.BASIC
        warnings.append("Plik bez koperty - przyjeto wartosci domyslne.")
    elif isinstance(payload, dict):
        declared = payload.get("format")
        if declared and declared != FORMAT_ID:
            warnings.append(f"Plik deklaruje format '{declared}', oczekiwano '{FORMAT_ID}'.")
        notes = payload.get("notes")
        if not isinstance(notes, list):
            raise ImportError_("Brak listy 'notes'.")
        deck_name = payload.get("deck")
        default_note_type = _note_type(payload.get("default_note_type"), warnings)
    else:
        raise ImportError_("Oczekiwano obiektu albo listy na najwyzszym poziomie.")

    rows: list[SourceRow] = []
    for index, note in enumerate(notes, start=1):
        if not isinstance(note, dict):
            warnings.append(f"Pozycja {index}: pominieta (nie jest obiektem).")
            continue

        values = {column: _as_text(note.get(key)) for key, column in _KEYS.items()}
        if not values[FIELD_FRONT].strip() or not values[FIELD_BACK].strip():
            warnings.append(f"Pozycja {index}: pominieta (pusty przod albo tyl).")
            continue

        tags = note.get("tags")
        rows.append(
            SourceRow(
                values=values,
                source_ref=_as_text(note.get("source_ref")) or None,
                source_deck=_as_text(note.get("deck")) or deck_name,
                tags=[str(t) for t in tags] if isinstance(tags, list) else [],
            )
        )

    if not rows:
        raise ImportError_("Plik nie zawiera zadnej kompletnej fiszki.")

    return ParseResult(
        source_format="fiszki-json",
        columns=list(_KEYS.values()),
        rows=rows,
        # Format kanoniczny nie wymaga zgadywania - kolumny sa juz docelowe.
        suggested_mapping={column: column for column in _KEYS.values()},
        warnings=warnings,
        source_decks=sorted({r.source_deck for r in rows if r.source_deck}),
        suggested_note_type=default_note_type,
    )


def _note_type(value: Any, warnings: list[str]) -> NoteType:
    if value is None:
        return NoteType.BASIC
    try:
        return NoteType(str(value))
    except ValueError:
        warnings.append(f"Nieznany typ notatki '{value}' - uzyto jednostronnego.")
        return NoteType.BASIC

File: frontend/src/lib/local/import/canonical.ts
/**
 * Format wlasny aplikacji - `fiszki/v1` (specyfikacja: docs/format-fiszki-v1.md).
 *
 * Format docelowy: kolumny sa juz docelowe, wiec import przechodzi bez
 * mapowania. Najwygodniejszy sposob dostarczenia gotowej talii z zewnatrz.
 */

import type { NoteType } from "@/lib/types";

import { FIELD_BACK, FIELD_EXAMPLE, FIELD_FRONT } from "../content";
import {
  ImportParseError,
  TARGET_KIND,
  TARGET_TAGS,
  type ParseResult,
  type SourceRow,
  sourceRow,
} from "./base";

export const FORMAT_ID = "fiszki/v1";

//: Klucz w JSON -> kolumna posrednia (nazwa widoczna pozniej w mapowaniu).
const KEYS: Record<string, string> = {
  front: FIELD_FRONT,
  back: FIELD_BACK,
  example: FIELD_EXAMPLE,
  tags: TARGET_TAGS,
  kind: TARGET_KIND,
};

function asText(value: unknown): string {
  if (value == null) return "";
  if (Array.isArray(value)) return value.map(String).join(", ");
  return String(value);
}

const NOTE_TYPES: NoteType[] = ["basic", "basic_reversed"];

function envelopeNoteType(value: unknown, warnings: string[]): NoteType {
  if (value == null) return "basic";
  if (NOTE_TYPES.includes(value as NoteType)) return value as NoteType;
  warnings.push(`Nieznany typ notatki '${String(value)}' - uzyto jednostronnego.`);
  return "basic";
}

/**
 * Typ notatki podany przy pojedynczej pozycji. null oznacza "uzyj typu
 * wybranego przy imporcie" - inaczej niz w kopercie, gdzie brak wartosci
 * oznacza wprost typ jednostronny.
 */
function itemNoteType(value: unknown, index: number, warnings: string[]): NoteType | null {
  if (value == null || !String(value).trim()) return null;
  const candidate = String(value).trim();
  if (NOTE_TYPES.includes(candidate as NoteType)) return candidate as NoteType;
  warnings.push(`Pozycja ${index}: nieznany typ notatki '${candidate}' - uzyto typu z importu.`);
  return null;
}

export function parse(data: Uint8Array): ParseResult {
  let payload: unknown;
  try {
    payload = JSON.parse(new TextDecoder("utf-8").decode(data));
  } catch (caught) {
    throw new ImportParseError(`Nieprawidlowy JSON: ${(caught as Error).message}`);
  }

  const warnings: string[] = [];
  let notes: unknown[];
  let deckName: string | null = null;
  let defaultNoteType: NoteType = "basic";

  if (Array.isArray(payload)) {
    notes = payload;
    warnings.push("Plik bez koperty - przyjeto wartosci domyslne.");
  } else if (payload && typeof payload === "object") {
    const envelope = payload as Record<string, unknown>;
    const declared = envelope.format;
    if (declared && declared !== FORMAT_ID) {
      warnings.push(`Plik deklaruje format '${String(declared)}', oczekiwano '${FORMAT_ID}'.`);
    }
    if (!Array.isArray(envelope.notes)) {
      throw new ImportParseError("Brak listy 'notes'.");
    }
    notes = envelope.notes;
    deckName = asText(envelope.deck) || null;
    defaultNoteType = envelopeNoteType(envelope.default_note_type, warnings);
  } else {
    throw new ImportParseError("Oczekiwano obiektu albo listy na najwyzszym poziomie.");
  }

  const rows: SourceRow[] = [];
  notes.forEach((note, i) => {
    const index = i + 1;
    if (!note || typeof note !== "object" || Array.isArray(note)) {
      warnings.push(`Pozycja ${index}: pominieta (nie jest obiektem).`);
      return;
    }
    const item = note as Record<string, unknown>;

    const values = Object.fromEntries(
      Object.entries(KEYS).map(([key, column]) => [column, asText(item[key])]),
    );
    if (!values[FIELD_FRONT].trim() || !values[FIELD_BACK].trim()) {
      warnings.push(`Pozycja ${index}: pominieta (pusty przod albo tyl).`);
      return;
    }

    rows.push(
      sourceRow({
        values,
        sourceRef: asText(item.source_ref) || null,
        sourceDeck: asText(item.deck) || deckName,
        tags: Array.isArray(item.tags) ? item.tags.map(String) : [],
        noteType: itemNoteType(item.note_type, index, warnings),
      }),
    );
  });

  if (rows.length === 0) {
    throw new ImportParseError("Plik nie zawiera zadnej kompletnej fiszki.");
  }

  return {
    sourceFormat: "fiszki-json",
    columns: Object.values(KEYS),
    rows,
    // Format kanoniczny nie wymaga zgadywania - kolumny sa juz docelowe.
    suggestedMapping: Object.fromEntries(Object.values(KEYS).map((c) => [c, c])),
    warnings,
    sourceDecks: [...new Set(rows.map((r) => r.sourceDeck).filter((d): d is string => !!d))].sort(),
    suggestedNoteType: defaultNoteType,
  };
}
